metrictank-sample.ini

```diff
@@ -19,9 +19,11 @@ metric-max-stale = 3h
 # Interval to run garbage collection job
 gc-interval = 1h
 
-# duration before secondary nodes start serving requests
-# shorter warmup means metrictank will need to query cassandra more if it doesn't have requested data yet.
-# in clusters, best to assure the primary has saved all the data that a newly warmup instance will need to query, to prevent gaps in charts
+# duration before non-primary (secondary or query) nodes start serving requests
```

Comment:
It doesnt make sense to me to have warm-up period applied to query nodes.  If waiting for gossip to settle is needed for query nodes, then it would also be needed for primary nodes, in which case warm-up-period should always apply.

The intent of the warm-up-period is to set the amount of time a node should spend rebuilding in-memory chunks.  If we need to set a time to wait for gossip to settle, we should add a new config setting.

Reply:
> If waiting for gossip to settle is needed for query nodes, then it would also be needed for primary nodes, in which case warm-up-period should always apply.

Let me explain where I'm coming from.

First of all with "gossip settle" I mean "become aware of all other cluster peers" or at the very least "become aware of at least one ready replica of each shard" (assuming a healthy cluster wherein all shards are ready)

Here's what happens during startup:
1) gossip start and http server coming online happens basically at the same time.
2) Then we load index, start indexpersist notifiers, start data consumption
3) now we set ready state (currently: immediately for primaries, after warmup otherwise)

My position is that we should not speculate that step 2 will always take some time.
In most cases, it'll be a cluster with already some data, and it'll take some time until we reach step 3, enough time for gossip to settle.
But, the cluster may only contain very little data and step 2 may go very quickly, or maybe the network is slow and gossip is taking some time. or it's an overprovisioned cluster. or peers are slow to gossip because they're overload, or whatever.
The point is, we cannot guarantee that by the time we reach step 3, gossip has settled.
Thus, if the node receives render queries it may not be able to respond them properly (as completely as other nodes that have been up longer, would have. or it may error depending on `min-available-shards`)

So generally speaking (for sharded instances that receive queries), I think we need to make sure to leave some time for gossip settle.
The argument that we shouldn't piggyback this on the warm-up-period setting, but instead introduce a 2nd variable for it, is fine with me. I have no strong preference either way.

As far as why do primaries become ready immediately at step 3, IIRC I think this is based on some old logic that if you're querying a primary that just started, you probably have no other instances online that serve the same data as this primary is holding. (otherwise why would you query the primary?) so may as well try our best and serve what we can. This works well in our deployments where we have primaries that we don't query. But for clusters wherein you query your secondaries and primaries, I think it would make sense to subject the primary to the same warm-up-period.

Reply:
after your explanation it is even more clear that time for gossip to settle needs to be a separate config setting.

in a cluster, nodes have 3 functions. A single MT node can provide one, two or all three.
a) ingesting and storing data in memory
b) writing data to backend store
c) executing queries.

for "a" and "b", gossip has nothing to do with whether the node is "ready".   Readiness is determined solely by the current state of the node. A node should not be marked as ready unless the union of data in-memory and data in the backend store is the complete set of data.  This can be ensured by
  - replaying from kafka
  - waiting a fixed "warm-up-period" to ensure that for any data older then what a node has in-memory, the "primary" node has written it to the backend-store.  The "warm-up-period" is skipped If a node is a "primary" because the data it has in-memory + what is in the store is already the full set of all data.

"c" (executing queries) is the only function that relies on gossip as it needs to know which other shards exist, and which nodes are "ready" for each shard.

Reply:
> The "warm-up-period" is skipped If a node is a "primary" because the data it has in-memory + what is in the store is already the full set of all data.

It depends: when not backfilling from kafka, i.e. with offset=newest , if a primary just restarted and we skip warm-up, recent data won't have been written yet to the store and won't be consumed from kafka either.
Hence my earlier point that we may want to subject primaries to warmup period as well. Though this is out of scope for this PR.

> "c" (executing queries) is the only function that relies on gossip as it needs to know which other shards exist, and which nodes are "ready" for each shard.

yes

> after your explanation it is even more clear that time for gossip to settle needs to be a separate config setting.

OK. I will add a new setting `cluster.gossip-settle-period` or something like that (note: should probably not go into the `swim` section because those all directly correspond to memberlist config settings) instead of abusing warm-up-period